netebpfext/net_ebpf_ext_sock_addr.c

```diff
@@ -0,0 +1,416 @@
+    uint16_t destination_port_field =
+        (direction == EBPF_HOOK_SOCK_ADDR_EGRESS) ? fields->remote_port_field : fields->local_port_field;
+
+    FWPS_INCOMING_VALUE0* incoming_values = incoming_fixed_values->incomingValue;
+
+    // Copy IP address fields.
+    if ((hook_id == EBPF_HOOK_SOCK_ADDR_CONNECT_V4) || (hook_id == EBPF_HOOK_SOCK_ADDR_RECV_ACCEPT_V4)) {
+        sock_addr_ctx->family = AF_INET;
+        sock_addr_ctx->msg_src_ip4 = incoming_values[source_ip_address_field].value.uint32;
+        sock_addr_ctx->user_ip4 = incoming_values[destination_ip_address_field].value.uint32;
+    } else {
+        sock_addr_ctx->family = AF_INET6;
+        RtlCopyMemory(
+            sock_addr_ctx->msg_src_ip6,
+            incoming_values[source_ip_address_field].value.byteArray16,
+            sizeof(FWP_BYTE_ARRAY16));
+        RtlCopyMemory(
+            sock_addr_ctx->user_ip6,
+            incoming_values[destination_ip_address_field].value.byteArray16,
+            sizeof(FWP_BYTE_ARRAY16));
+    }
+    sock_addr_ctx->msg_src_port = incoming_values[source_port_field].value.uint16;
+    sock_addr_ctx->user_port = incoming_values[destination_port_field].value.uint16;
+    sock_addr_ctx->protocol = incoming_values[fields->protocol_field].value.uint8;
+    sock_addr_ctx->compartment_id = incoming_values[fields->compartment_id_field].value.uint32;
+}
+
+//
+// WFP callout callback function.
+//
+void
+net_ebpf_extension_sock_addr_authorize_connection_classify(
+    _In_ const FWPS_INCOMING_VALUES* incoming_fixed_values,
+    _In_ const FWPS_INCOMING_METADATA_VALUES* incoming_metadata_values,
+    _Inout_opt_ void* layer_data,
+    _In_opt_ const void* classify_context,
+    _In_ const FWPS_FILTER* filter,
+    uint64_t flow_context,
+    _Inout_ FWPS_CLASSIFY_OUT* classify_output)
+{
+    uint32_t result;
+    net_ebpf_extension_hook_client_t* attached_client = NULL;
+    net_ebpf_extension_hook_execution_t execution_type =
+        (KeGetCurrentIrql() < DISPATCH_LEVEL) ? EXECUTION_PASSIVE : EXECUTION_DISPATCH;
+    FWP_ACTION_TYPE action = FWP_ACTION_PERMIT;
+    bpf_sock_addr_t sock_addr_ctx = {0};
+
+    UNREFERENCED_PARAMETER(incoming_metadata_values);
+    UNREFERENCED_PARAMETER(layer_data);
+    UNREFERENCED_PARAMETER(classify_context);
+    UNREFERENCED_PARAMETER(flow_context);
+
+    attached_client = (net_ebpf_extension_hook_client_t*)filter->context;
+    ASSERT(attached_client != NULL);
+    if (attached_client == NULL)
+        goto Exit;
+
+    if (!net_ebpf_extension_hook_client_enter_rundown(attached_client, execution_type)) {
+        classify_output->actionType = FWP_ACTION_PERMIT;
+        goto Exit;
+    }
+
+    _net_ebpf_extension_sock_addr_copy_wfp_connection_fields(incoming_fixed_values, &sock_addr_ctx);
+
+    if (net_ebpf_extension_hook_invoke_program(attached_client, &sock_addr_ctx, &result) != EBPF_SUCCESS)
+        goto Exit;
+
+    action = (result == VERDICT_PROCEED) ? FWP_ACTION_PERMIT : FWP_ACTION_BLOCK;
```

Comment:
(just an observation that this treats any other value as block, so it need not be negative as the comments in the header said)